ush/python/pygw/src/pygw/executable.py

```diff
@@ -0,0 +1,315 @@
+                used (envisaged for testing purposes only)
+            env (dict): The environment with which to run the executable
+            fail_on_error (bool): Raise an exception if the subprocess returns
+                an error. Default is True. The return code is available as
+                ``exe.returncode``
+            ignore_errors (int or list): A list of error codes to ignore.
+                If these codes are returned, this process will not raise
+                an exception even if ``fail_on_error`` is set to ``True``
+            input: Where to read stdin from
+            output: Where to send stdout
+            error: Where to send stderr
+
+        Accepted values for input, output, and error:
+
+        * python streams, e.g. open Python file objects, or ``os.devnull``
+        * filenames, which will be automatically opened for writing
+        * ``str``, as in the Python string type. If you set these to ``str``,
+          output and error will be written to pipes and returned as a string.
+          If both ``output`` and ``error`` are set to ``str``, then one string
+          is returned containing output concatenated with error. Not valid
+          for ``input``
+        * ``str.split``, as in the ``split`` method of the Python string type.
+          Behaves the same as ``str``, except that value is also written to
+          ``stdout`` or ``stderr``.
+
+        By default, the subprocess inherits the parent's file descriptors.
+
+        """
+        # Environment
+        env_arg = kwargs.get("env", None)
+
+        # Setup default environment
+        env = os.environ.copy() if env_arg is None else {}
+        env.update(self.default_env)
+
+        # Apply env argument
+        if env_arg:
+            env.update(env_arg)
+
+        if "_dump_env" in kwargs:
+            kwargs["_dump_env"].clear()
+            kwargs["_dump_env"].update(env)
+
+        fail_on_error = kwargs.pop("fail_on_error", True)
+        ignore_errors = kwargs.pop("ignore_errors", ())
+
+        # If they just want to ignore one error code, make it a tuple.
+        if isinstance(ignore_errors, int):
+            ignore_errors = (ignore_errors,)
+
+        output = kwargs.pop("output", None)
+        error = kwargs.pop("error", None)
+        input = kwargs.pop("input", None)
+
+        if input is str:
+            raise ValueError("Cannot use `str` as input stream.")
+
+        def streamify(arg, mode):
+            if isinstance(arg, str):
+                return open(arg, mode), True
+            elif arg in (str, str.split):
+                return subprocess.PIPE, False
+            else:
+                return arg, False
+
+        istream, close_istream = streamify(input, "r")
+        ostream, close_ostream = streamify(output, "w")
+        estream, close_estream = streamify(error, "w")
+
+        cmd = self.exe + list(args)
+
+        escaped_cmd = ["'%s'" % arg.replace("'", "'\"'\"'") for arg in cmd]
```

Comment:
I think this should be functionally identical:

```suggestion
        escaped_cmd = [f"'{arg.replace("'", "'\"'\"'")}'" for arg in cmd]
```

Reply:
If you are sure, I can commit.  I am a bit leary of all the `'` and `"`.  They play tricks with my eyes.

Reply:
I used the same quotation marks as were there, I just replaced `%s` with the substitution. Let me check real fast.

Reply:
It does not work. Just leave as-is for now, as much as using `%` pains me.